### DISCUSSION TITLE
`Doc<T>` in `change` function callbacks

### DISCUSSION THREAD

**Discussion by acurrieclark** [automerge/automerge#1517] [General]
Following on from the `Extend` purge (RIP), I wanted to touch on the use of `Doc<T>` when used within `change` function callbacks.

`Doc<T>` is a really useful description of a document outside of a change function; It is readonly and should not be changed. 

However, when you are inside a change function, as far as the consumer is concerned, they are dealing with `T`. This is the type passed to the `change` function callback.

The real issues are with the functions which are only used within the change callback. For if we start with a simple document:

```ts
const doc = from({
  words: "are in here"
});
```
then this change is valid, and will work:
```ts
change(doc, (d) => {
  splice(d, ["words"], 0, 3, "look good")';
});
```
However, this change will not show any kind of typescript error but will throw a runtime error as `d.words` is not an object
```ts
change(doc, (d) => {
  splice(d.words, [], 0, 3, "look good")';
});
```
Do we need to narrow the type? Perhaps `splice` should just work in this scenario so we could update the function to accept the object we want to splice directly?

Within `splice`, `doc` is assumed to be `Doc<T>`, and most of the internal functions take `Doc<T>` as a parameter, so this isn't straightforward either way.

**Comment by alexjg** [comment]
Is the typing of `splice`, `insertAt`, and `deleteAt` the only problem here? I.e. the problem would be solved with the use of the `Path<T>` type you discussed in #630 ?

**Comment by acurrieclark** [comment]
In a roundabout way, maybe. At the moment you could technically (as far as typescript is concerned) pass a string in as the first parameter, which wouldn't have a path. This wouldn't error right now either:

```ts
change(doc, (d) => {
  splice("some-random-string", ["words"], 0, 3, "look good")';
});
```

Obviously this fails at runtime. Using `Path<T>` would require `T` to extend an object at least.

**Comment by alexjg** [comment]
Ah you're saying that splice should be `splice<T>(doc: T, ...)` rather than `splice<T>(doc: Doc<T>, ...)`?

**Comment by acurrieclark** [comment]
Yes, pretty much. Then you could constrain T. Something like:

```ts
function splice<
  const T extends Record<PropertyKey, any>, 
  TPath extends Path<T>
>(doc: T, path: TPath<T>, etc...) {}
```

You would then need to do lots of `as Doc<T>` within the splice function, as lots of the functions require that as an argument type.

**Comment by mweidner037** [comment]
Possibly related: I'm writing a React component that takes a `changeDoc` function as a prop (expecting it to come from a `useDocument` hook). If I type the prop as
```ts
// import { next as A } from "@automerge/automerge";

changeDoc: (
  changeFn: A.ChangeFn<unknown>,
  options?: A.ChangeOptions<unknown> | undefined
) => void;
```
then `splice` is unhappy when I call it inside `changeDoc`:
```ts
changeDoc((doc) =>
  A.splice(doc, path, ...)
  // ts(2345): Argument of type "unknown" is not assignable to parameter of type 'Doc<unknown>'
);
```
